### PR TITLE
Implement po_create()/po_update() for creating/updating translations

### PR DESCRIPTION
**Pull request by hadley**
I don't love the name `tr_add()`, and it's making me doubt that the developer/translator distinction is that useful, because this function might be run by the developer or an R-packages savvy translator.

I tweaked the `.Rproj` to make my usual roxygen2 workflow a little smoother; let me know if you want to revert.

I'm little surprised that potools doesn't already use `msginit`, so maybe I should instead copy the approach used in `translate_package()`?

### CODE REVIEW

File: R/msgmerge.R [hunk 1]
```diff
@@ -1,6 +1,15 @@
 # split off from tools::update_pkg_po() to only run the msgmerge & checkPoFile steps
+
+# https://www.gnu.org/software/gettext/manual/html_node/msgmerge-Invocation.html
 run_msgmerge = function(po_file, pot_file) {
-  if (system(sprintf("msgmerge --update %s %s", po_file, shQuote(pot_file))) != 0L) {
+  cmd <- paste("msgmerge",
+    "--update",
+    "--previous", # show previous match for fuzzy matches
```
Comment (hadley) [inline R/msgmerge.R:7]: This makes fuzzy matches considerably more useful because it shows you the old message and the new message side-by-side

**Comment by hadley** [comment]
Changed it to add or update, depending on whether or not the `.po` file already exists. That makes `tr_add()` and even worse name.

**Comment by hadley** [comment]
I think this is better as `po_create()` and `po_update()`.  I think `po_update()` should update all languages if none supplied; but that needs code from #234, so I'll wait until you've have a chance to review that.

File: potools.Rproj [hunk 2]
```diff
@@ -18,3 +18,4 @@ StripTrailingWhitespace: Yes
 BuildType: Package
 PackageUseDevtools: Yes
 PackageInstallArgs: --no-multiarch --with-keep.source
+PackageRoxygenize: rd
```
Comment (MichaelChirico) [inline potools.Rproj:21]: Could you elaborate what this does? Happy to add it, just don't see anything in a quick search

**Comment by MichaelChirico** [comment]
> I'm little surprised that potools doesn't already use msginit

There are a couple of gettext tools I would do well to integrate better with / provide good wrappers for...

File: R/msgmerge.R [hunk 3]
```diff
@@ -70,3 +79,66 @@ update_en_quot_mo_files <- function(dir, verbose) {
...
+#'
+#' @description
+#' * `po_create()` creates a new `po/{lang}.po` containing the messages to be
+#'   translated.
+#' * `po_update()` updates an existing `.po` file after the messages in a
+#'   package have changed. The translations for existing messages are preserved;
+#'   new messages are added; and translations for deleted message are marked
+#'   as deprecated and moved to the bottom of the file.
+#'
+#' @param lang Language identifiers. These are typically two letters (e.g.
```
Comment (MichaelChirico) [inline R/msgmerge.R:95]: This basically repeats the description in `?translate_package`... `@inheritParams` doesn't feel quite right because of how detailed we want to be... Maybe we need a `?language-code` .Rd file we can refer to? `@param languages Language identifiers. See [language-code] for details.`?

File: R/msgmerge.R [hunk 4]
```diff
@@ -70,3 +79,66 @@ update_en_quot_mo_files <- function(dir, verbose) {
...
+#'
+#' @description
+#' * `po_create()` creates a new `po/{languages}.po` containing the messages to be
+#'   translated.
+#' * `po_update()` updates an existing `.po` file after the messages in a
+#'   package have changed. The translations for existing messages are preserved;
+#'   new messages are added; and translations for deleted message are marked
+#'   as deprecated and moved to the bottom of the file.
+#'
+#' @param languages Language identifiers. These are typically two letters (e.g.
```
Comment (MichaelChirico) [inline R/msgmerge.R:95]: Wording here implies it's vectorized but we use `file.exists(po_path)` below... which do we want to fix? I would guess vectorization is better to support

File: R/msgmerge.R [hunk 5]
```diff
@@ -70,3 +79,66 @@ update_en_quot_mo_files <- function(dir, verbose) {
...
+#' @rdname po_create
+po_update <- function(languages, dir = ".", verbose = TRUE) {
+  package <- get_desc_data(dir)[["Package"]]
+
+  if (verbose) messagef("Updating existing %s translation", languages)
+  run_msgmerge(po_path(dir, languages), pot_path(dir, package), previous = TRUE)
+}
+
+po_path <- function(dir, lang) {
+  file.path(dir, "po", paste0("R-", lang, ".po"))
```
Comment (MichaelChirico) [inline R/msgmerge.R:124]: what about src translations?

File: potools.Rproj (hunk 2 above)
Comment (hadley) [reply potools.Rproj:21]: This just tell RStudio to use roxygen2 only for .Rd files and not the namespace.

File: R/msgmerge.R (hunk 4 above)
Comment (hadley) [reply R/msgmerge.R:95]: I've played around with this a bit, and I think it'll be easiest to implement simultaneously with adding support for src translations (oops!) after #234 provides `po_files()` (or whatever it gets called).

**Comment by hadley** [comment]
Ok, reimplemented basically from scratch; I'll add a bunch of tests shortly.

File: tests/testthat/_snaps/po_create.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+      Running msgint on './po/R-jp.po' succeeded:
+        Created ./po/R-jp.po.
+
+---
+
+    Code
+      po_create("jp", verbose = TRUE)
+    Message <simpleMessage>
+      Updating 'jp' R translation
+      Running msgmerge on './po/R-jp.po' succeeded:
```
Comment (hadley) [inline tests/testthat/_snaps/po_create.md:16]: I copied this from #257, but it feels a bit wordy to me. What do you think about not showing the call by default?

File: R/write_po_file.R [hunk 7]
```diff
@@ -300,7 +300,7 @@ format.po_metadata = function(x, template = FALSE, use_plurals = FALSE, ...) {
     x$email = "EMAIL@ADDRESS"
     x$language = ''
     x$language_team = "LANGUAGE <LL@li.org>"
-    x$charset = 'CHARSET'
+    x$charset = 'UTF-8'
```
Comment (hadley) [inline R/write_po_file.R:303]: I'm not sure why this was CHARSET before, but I was getting an error from `tools::checkPoFile()` that it was failing to `iconv` `CHARSET` to `UTF-8`.

File: R/msgmerge.R [hunk 8]
```diff
@@ -1,16 +1,23 @@
...
     # nocov these warnings? i don't know how to trigger them as of this writing.
-    warningf("Running msgmerge on '%s' failed.", po_file)
+    warningf("Running msgmerge on './po/%s' failed:\n  %s", basename(po_file), paste(val, collapse = "\n"))
+  } else if (verbose) {
+    messagef("Running msgmerge on './po/%s' succeeded:\n  %s", basename(po_file), paste(val, collapse = "\n"))
   }
 
-  res <- tools::checkPoFile(po_file, strictPlural = TRUE)
-  if (nrow(res)) {
-    warningf("tools::checkPoFile() found some issues in %s", po_file)
```
Comment (MichaelChirico) [inline R/msgmerge.R:10]: any reason to remove this warning? IINM `tools::checkPoFile()` doesn't have any incidental effect so this call will now do nothing

File: R/write_po_file.R (hunk 7 above)
Comment (MichaelChirico) [reply R/write_po_file.R:303]: Hmm note that .pot is not mentioned in `?tools::checkPoFile`. `CHARSET` shows up in templates in base:

https://github.com/wch/r-source/blob/bf5885960808d7c539e74b238575f54de0faba04/src/library/base/po/R.pot#L17

https://github.com/wch/r-source/blob/bf5885960808d7c539e74b238575f54de0faba04/src/library/base/po/R-base.pot#L10

Here's what I'm seeing in the `xgettext` docs:

https://www.gnu.org/software/gettext/manual/gettext.html

> `MIME-Version, Content-Type, Content-Transfer-Encoding`
> These values are set according to the content of the POT file and the current locale. If the POT file contains charset=UTF-8, it means that the POT file contains non-ASCII characters, and we keep the UTF-8 encoding. Otherwise, when the POT file is plain ASCII, we use the locale’s encoding.

and 

> `Content-Type`
> Replace ‘CHARSET’ with the character encoding used for your language, in your locale, or UTF-8. This field is needed for correct operation of the msgmerge and msgfmt programs, as well as for users whose locale’s character encoding differs from yours (see Charset conversion).

In the gettext sources, here is where `xgettext` initially populates the default `CHARSET`:

https://github.com/autotools-mirror/gettext/blob/030c0341a4a0ba6ad7fe62e83ff663bdc76cbe4d/gettext-tools/src/xgettext.c#L2015-L2028

Here `write-po` treats `CHARSET` as ASCII:

https://github.com/autotools-mirror/gettext/blob/030c0341a4a0ba6ad7fe62e83ff663bdc76cbe4d/gettext-tools/src/write-po.c#L1650-L1653

This snippet makes me thing setting UTF-8 is probably OK, since the default `CHARSET` has the probably-in-general faulty assumption that the strings in the .pot file are all-ASCII:

https://github.com/autotools-mirror/gettext/blob/030c0341a4a0ba6ad7fe62e83ff663bdc76cbe4d/gettext-tools/src/po-charset.c#L490-L509

That said, current r-devel .pot are all-ASCII:

```
grep -rP "[\x80-\xFF]" ~/svn/R-devel/ --include=*.pot
# <empty>
```

CRAN would require this to hold true for all packages at the R level, and probably the C level?

Comment (MichaelChirico) [reply R/write_po_file.R:303]: We could probably patch `tools::checkPoTools` to accept the `CHARSET` :arrow_right: `ASCII` logic, but I'm not sure that would solve all of the issues of using `checkPoTools()` on .pot files when it was designed for .po.

File: tests/testthat/_snaps/po_create.md (hunk 6 above)
Comment (MichaelChirico) [reply tests/testthat/_snaps/po_create.md:16]: Hmm it's coming from `verbose=TRUE` (now the default), are you suggesting we flip the default, or that we implement `verbose=0,1,2,...`?

File: R/utils.R [hunk 9]
```diff
@@ -167,3 +167,19 @@ is_outdated <- function(src, dst) {
 }
 
 is_testing = function() identical(Sys.getenv("TESTTHAT"), "true")
+
+local_test_package <- function(..., .envir = parent.frame()) {
```
Comment (MichaelChirico) [inline R/utils.R:171]: love the look of this, and looks like a great redesign of generating test packages (the current suite is pretty unwieldy/arbitrarily organized). a good companion would be `local_translation_conn()` to keep translations right next to where they'll be used too.

but feels like it should be in `tests/testthat/helper.R`, no?

File: R/write_po_file.R (hunk 7 above)
Comment (hadley) [reply R/write_po_file.R:303]: Maybe I'll back this change out here and we can discuss more thoroughly in a separate issue. I think potools should assume all `.pot` and `.po` files are UTF-8 and then fix any downstream problems that causes.

File: tests/testthat/_snaps/po_create.md (hunk 6 above)
Comment (hadley) [reply tests/testthat/_snaps/po_create.md:16]: I don't think it's necessary to routinely display a message for success. Adding levels of verbosity is unlikely to be worth the effort IMO, so I'd be in favour of `verbose = FALSE` by default and not passing in `verbose` from the wrapping function.

File: R/write_po_file.R (hunk 7 above)
Comment (hadley) [reply R/write_po_file.R:303]: Hmmm, backing out this change doesn't break any of my tests (or interactive usage) so definitely doesn't need to be here.

File: tests/testthat/_snaps/po_create.md (hunk 6 above)
Comment (hadley) [reply tests/testthat/_snaps/po_create.md:16]: Oh that's not quite right; I think on success we want to see the stdout/stderr from the command line tool, but not the call.

Comment (hadley) [reply tests/testthat/_snaps/po_create.md]: I've tweaked the code to behave the way that I want just to give you a concrete proposal to look at. Let me know what you think.

File: R/po_create.R [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    package <- get_desc_data(dir, "Package")
+  }
+  if (length(type) == 0) {
+    character()
+  } else {
+    file.path(dir, "po", paste0(po_prefix(type), package, ".pot"))
+  }
+
+}
+po_prefix <- function(type = c("R", "src")) {
```
Comment (MichaelChirico) [inline R/po_create.R:55]: the one place this will break is for `base`, which uses `R-base.pot` for R and `R.pot` for src. I'll just add a TODO for now to make sure that case is handled later.

File: R/po_create.R [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+#'   "en" = English, "fr" = French, "es" = Spanish, "zh" = Chinese), but
+#'   can include an additional suffix for languages that have regional
+#'   variations (e.g. "fr_CN" = French Canadian, "zh_CN" = simplified
+#'   characters as used in mainland China, "zh_TW" = traditional characters
+#'   as used in Taiwan.)
+#' @inheritParams po_extract
+#' @export
+po_create <- function(languages, dir = ".", verbose = !is_testing()) {
+  package <- get_desc_data(dir, "Package")
+  po_files <- po_language_files(languages, dir)
```
Comment (MichaelChirico) [inline R/po_create.R:21]: it might be nice to `warn()` here if e.g. R-pkg.pot is present but pkg.pot is not & the package has a `src/` directory.

OTOH, if the translation is handled entirely by `potools`, this should only happen if `src/` doesn't have any messages to translate, right? So maybe it will be a warning highly prone to false positives...

skipping for now, but file a follow-up issue if you think the warning is worthwhile.

**Comment by MichaelChirico** [comment]
Looking great, thanks!
